### PR TITLE
Fix #113

### LINKED ISSUE #113: Convert Toolbox from XML to JSON
## Reason

https://developers.google.com/blockly/guides/configure/web/serialization:
  > Blockly provides two formats for this data: JSON and XML. We recommend using the JSON system for new projects, and encourage older projects using XML to upgrade. The XML system is the legacy save format. It will not be removed, but it will not receive new features.

## Subject

- **File** : <https://github.com/scratch-for-discord/Web-Application_Frontend/blob/release/2/master/src/toolbox.ts>

## Links

- **[npm package](https://github.com/scratch-for-discord/Web-Application_Frontend/issues/113#issuecomment-1318389663)**

### PR DESCRIPTION
**Pull request by Malix-Labs**

### CODE REVIEW

File: src/toolbox.ts [hunk 1]
```diff
@@ -1,46 +1,136 @@
...
+                {
+                    "type": "env",
+                    "kind": "block"
+                }
+            ]
+        }
+    ]
+}
 
 export { toolboxString }
```
Comment (Malix-Labs) [inline src/toolbox.ts]:
```suggestion
export { toolboxString }

```

File: src/toolbox.ts [hunk 2]
```diff
@@ -1,46 +1,136 @@
...
-
-        <block type='on_bot_code_start'/>
-        <block type='bot_amount_info'/>
-        <block type="bot_shutdown_stop"/>
-        <block type="set_bot_game_stream"/>
-        <block type="env"/>
-        </category>
-
-        </xml>`
+// done in a seconds with https://replit.com/@RedMan13/blockly-xml-to-json-toolbox
```
Comment (Malix-Labs) [inline src/toolbox.ts]:
```suggestion
```

Comment (Malix-Labs) [reply src/toolbox.ts]: We could credit it in a contributor system file later

Comment (RedMan13) [reply src/toolbox.ts]: its litteraly just me XD

Comment (Malix-Labs) [reply src/toolbox.ts]: ye ikr but we won't do credits with comments, we will do that properly

Comment (RedMan13) [reply src/toolbox.ts]: ah which file is it? or is it not here yet

**Comment by Malix-Labs** [comment]
@RedMan13 have you tested the whole new toolbox? every single components of it?

**Comment by RedMan13** [comment]
thats what i remember

**Comment by Malix-Labs** [comment]
@RedMan13:
> thats what i remember

what do you mean?

**Comment by RedMan13** [comment]
idk
i need to add json serialization anyways so ill just test now

**Comment by Malix-Labs** [comment]
I didn't deleted it??? i renamed
let me fix that

**Comment by RedMan13** [comment]
> I didn't deleted it??? i renamed let me fix that

github doesnt know what a rename is and considers it a delete and an upload

**Comment by Malix-Labs** [comment]
@RedMan13 
> github doesnt know what a rename is and considers it a delete and an uploa

It's fixed, you can review.

**Comment by RedMan13** [comment]
oh my fucking god im dumb
the reason theres no json serialization is because theres no serialization at all

**Comment by Malix-Labs** [comment]
Also @RedMan13, does your changed change everything from XML to JSON, or just the toolbox?
How about blocks option saving, and the XML s4d save file download?

**Comment by RedMan13** [comment]
there is nothing but toolbox in it rn
literally there is no serialization/deserialization (no save/load) in there, its only the toolbox

**Comment by Malix-Labs** [comment]
- [Comment Reply](https://github.com/scratch-for-discord/Web-Application_Frontend/pull/164#issuecomment-1383673965) from @RedMan13:
  > there is nothing but toolbox in it rn
literally there is no serialization/deserialization (no save/load) in there, its only the toolbox

  Then, the name of the branch is kinda misleading.
  Also, can you also convert serialization from XML to JSON please?

File: src/toolbox.ts (hunk 2 above)
Comment (Malix-Labs) [reply src/toolbox.ts]: There is none yet, we'll see that for v2 to make it automatic, np

**Comment by RedMan13** [comment]
there is no serialization to convert

**Comment by Malix-Labs** [comment]
- [Comment Reply](https://github.com/scratch-for-discord/Web-Application_Frontend/pull/164#issuecomment-1383702922) from @RedMan13:
  > there is no serialization to convert

  How so?
  If we change toolbox to JSON, the serialization will be in JSON automatically?

**Comment by RedMan13** [comment]
> If we change toolbox to JSON, the serialization will be in JSON automatically?

yes

**Comment by RedMan13** [comment]
the way blockly works is that you are not telling it to use json, you are just handing or requesting from it in json and it just does it

**Comment by Malix-Labs** [comment]
Okay, then do you confirm that by merging #164, we will fully close #113?

**Comment by RedMan13** [comment]
uh leme see if theres anyway i can fix the gap issue rq

**Comment by Malix-Labs** [comment]
Ok np

**Comment by Malix-Labs** [comment]
We need this PR to convert XML to JSON so that the toolbox stays exactly identical while using JSON instead of XML.

File: src/App.vue [hunk 3]
```diff
@@ -60,7 +60,7 @@ const options = {
     minScale: 0.1,
     scaleSpeed: 1.2
   },
-  toolbox: toolboxString,
+  toolbox: toolbox,
```
Comment (Malix-Labs) [inline src/App.vue:63]: Would removing this line change anything?

Comment (RedMan13) [reply src/App.vue:63]: It would change everything since that's what hands the toolbox to blockly

File: src/App.vue [hunk 4]
```diff
@@ -4,7 +4,7 @@
 import Blockly from 'blockly/core';
 import { ref } from "vue";
 import BlocklyComponent from "./components/BlocklyComponent.vue";
-import { toolboxString } from './toolbox'
+import toolbox from './toolbox'
 import { javascriptGenerator } from "blockly/javascript";
```
Comment (Malix-Labs) [inline src/App.vue:8]: Could we normalize the way we import, or it is required that we put some specific ones into brackets? If so, why?

File: src/toolbox.ts [hunk 5]
```diff
@@ -130,6 +126,4 @@ let toolboxString = {
             ]
         }
     ]
-}
-
-export { toolboxString }
+}
```
Comment (Malix-Labs) [inline src/toolbox.ts]:
```suggestion
}

```

File: src/App.vue (hunk 3 above)
Comment (Malix-Labs) [reply src/App.vue:63]: Okay

File: src/App.vue (hunk 4 above)
Comment (RedMan13) [reply src/App.vue:8]: the brackets extract the value from the import, they are technically pointless but they minimize the usage of `module.thing.thing.thing` when using larger modules with lots of functions and values and various different places

Comment (Malix-Labs) [reply src/App.vue:8]: > the brackets ⋯ are technically **pointless** but **they minimize the usage of module.thing.thing.thing when using larger modules with lots of functions and values and various different places**

Then, they are not really pointless

Comment (RedMan13) [reply src/App.vue:8]: yeah
i removed them from toolbox value because for that they litteraly where pointless

**Comment by Malix-Labs** [comment]
As this PR is not ready to be merged, please convert it to draft

**Comment by RedMan13** [comment]
> As this PR is not ready to be merged, please convert it to draft

it is ready though?

**Comment by RedMan13** [comment]
what about it isnt ready

**Comment by Malix-Labs** [comment]
> what about it isnt ready

Converter fixing and change on implementation

**Comment by RedMan13** [comment]
the converter is entirely unrelated to this

**Comment by RedMan13** [comment]
> @Malix-Off 
change on implementation

what?

**Comment by Malix-Labs** [comment]
You told me there were some stuff to fix about toolbox types and a "flyout" thingy, and that you would have to fix the converter.
If it's already fixed and you confirm this PR does close #113 as a whole by replacing everything XML to JSON, you can merge it.

**Comment by RedMan13** [comment]
> You told me there were some stuff to fix about toolbox types and a "flyout" thingy, and that you would have to fix the converter.

flyout and toolbox types are the same issue and where a instant fix, the converter is [xml to json](https://github.com/RedMan13/toolbox-xml-to-json)
> If it's already fixed and you confirm this PR does close https://github.com/scratch-for-discord/Web-Application_Frontend/issues/113 as a whole by replacing everything XML to JSON, you can merge it.

kk

**Pull request merged**